include/cuco/detail/static_set/kernels.cuh

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuco/detail/utils.hpp>
+
+#include <cub/block/block_reduce.cuh>
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+namespace cuco {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Inserts all elements in the range `[first, first + n)` and returns the number of
+ * successful insertions.
+ *
+ * If multiple elements in `[first, first + size)` compare equal, it is unspecified which
+ * element is inserted.
+ *
+ * @tparam BlockSize Number of threads in each block
+ * @tparam InputIterator Device accessible input iterator whose `value_type` is
+ * convertible to the set's `value_type`
+ * @tparam Reference Type of device reference allowing access to storage
+ *
+ * @param first Beginning of the sequence of input elements
+ * @param n Number of input elements
+ * @param num_successes Number of successful inserted elements
+ * @param set_ref Set device reference used to access the slot storage
+ */
+template <int32_t BlockSize, typename InputIterator, typename Reference>
+__global__ void insert(InputIterator first,
+                       cuco::detail::index_type n,
+                       typename Reference::size_type* num_successes,
+                       Reference set_ref)
+{
+  using BlockReduce = cub::BlockReduce<typename Reference::size_type, BlockSize>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  typename Reference::size_type thread_num_successes = 0;
+
+  cuco::detail::index_type const loop_stride = gridDim.x * BlockSize;
+  cuco::detail::index_type idx               = BlockSize * blockIdx.x + threadIdx.x;
+
+  while (idx < n) {
+    typename Reference::value_type const insert_pair{*(first + idx)};
+    if (set_ref.insert(insert_pair)) { thread_num_successes++; };
+    idx += loop_stride;
+  }
+
+  // compute number of successfully inserted elements for each block
+  // and atomically add to the grand total
+  typename Reference::size_type block_num_successes =
+    BlockReduce(temp_storage).Sum(thread_num_successes);
+  if (threadIdx.x == 0) {
+    cuda::atomic_ref<typename Reference::size_type, Reference::scope> count_ref(*num_successes);
+    count_ref.fetch_add(block_num_successes, cuda::std::memory_order_relaxed);
```

Comment:
We could use `atomicAdd` here to ship around that bug where `cuda::atomic_ref`'s ctor allocates local memory.

Reply:
Use `counter_storage` instead. Done.